From 5db240ec79eca1c5992d5a2c17e81d60e6cab63f Mon Sep 17 00:00:00 2001
From: Sultan Alsawaf <sultan@kerneltoast.com>
Date: Sat, 2 Dec 2023 20:58:01 -0800
Subject: [PATCH] cpuidle: Reject idle entry if need_resched() is true

There's no reason to enter idle at this point in __CPU_PM_CPU_IDLE_ENTER()
if the CPU needs to reschedule. Instead of fruitlessly entering the
architecture's idle routine, reject the idle entry attempt with an error as
an optimization.

Signed-off-by: Sultan Alsawaf <sultan@kerneltoast.com>
---
 include/linux/cpuidle.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

Index: common/include/linux/cpuidle.h
===================================================================
--- common/include/linux/cpuidle.h	2025-10-14 12:24:32.165691385 +1030
+++ common/include/linux/cpuidle.h	2025-10-14 12:24:32.161750409 +1030
@@ -316,7 +316,9 @@
 ({									\
 	int __ret = 0;							\
 									\
-	if (!idx) {							\
+	if (need_resched()) {						\
+		__ret = -1;						\
+	} else if (!idx) {						\
 		cpu_do_idle();						\
 		return idx;						\
 	}								\
